src/Core/src/Credentials/PersistingCacheProvider.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core\Credentials;
+
+use AsyncAws\Core\Configuration;
+use Symfony\Contracts\Cache\CacheInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+
+/**
+ * Cache the Credential generated by the decorated CredentialProvider in a persisting pool.
+ *
+ * The Credential will be reused until it expires.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+final class PersistingCacheProvider implements CredentialProvider
+{
+    private const EXPIRATION_DRIFT = 30;
+
+    private $cache;
+    private $decorated;
+
+    public function __construct(CredentialProvider $decorated, CacheInterface $cache)
+    {
+        $this->decorated = $decorated;
+        $this->cache = $cache;
+    }
+
+    public function getCredentials(Configuration $configuration): ?Credentials
+    {
+        return $this->cache->get(sha1(\serialize([$configuration, \get_class($this->decorated)])), function (ItemInterface $item) use ($configuration) {
+            $credential = $this->decorated->getCredentials($configuration);
+            if (null === $credential) {
+                $item->expiresAfter(0);
```

Comment:
This is also good to remember, right?
Or every user http request will make this request and get nothing back. 

Let’s save it for 5 seconds?

Reply:
I though like you,but then switch to 0:
In fact HTTP calls always return a credential. There are 2 cases where it return null:
- endpoint is not reachable (ie, the chain use InstanceCredential in a server that is not running in AWS)
- endpoint is temporary unreachable (network issue, when fetching credential, or timeout (remember 1.0 sec))

The first one should not happens in production. InstanceCredential is the last provider in the chain, and reaching it while the application in not runing in AWS means a wrong configuration.
The second one is temporary and should not be cached IMHO

Reply:
Okey, You are correct. Thank you